### ISSUE TITLE
PamResponse is not being properly initalised in FFI

### ISSUE THREAD

**Issue by carnott-snap** [rcatolino/pam_sm_rust#17]
Some code that was working 4m back, appears to now be breaking. Since nothing changed with this library, I assume it is a spec issue, as my Linux has upgraded a few times since. Specifically the call to `CStr::from_ptr(r)` fails because `r` is null:
https://github.com/rcatolino/pam_sm_rust/blob/3ac36345b6f9c67d6f482d0da59df489601221d1/src/pam.rs#L134

This leads to a segfault:
stdout:
```
Segmentation fault (core dumped)
```
dmesg:
```
pamtester[567743]: segfault at ... error 4 in libc-2.31.so[...]
```

**Comment by lucab** [comment]
@carnott-snap which libpam version is that using? Also, do you maybe have a small standalone example exercising this codepath?

**Comment by carnott-snap** [comment]
Version from aptitude:
`libpam0g        1.3.1-5ubuntu4  Pluggable Authentication Modules library`

This is the affected line, let me know if you need more:
```rust
let key = pamh.conv(Some("totp: "), PamMsgStyle::PROMPT_ECHO_ON)?.to_string_lossy();
```

**Comment by rcatolino** [comment]
The man page never says that the ```resp``` member of ```struct pam_response``` cannot be null, so I wouldn't call that a spec error.

I believe the issue comes from our definition of ```PamResponse``` : I thought that ```Option<*mut c_char>``` was guaranteed to map to ```None``` when ```*c_char``` is null, but reading [the rustonomicon page on this](https://doc.rust-lang.org/nomicon/ffi.html#the-nullable-pointer-optimization), it's not so clear : 

>As a special case, an enum is eligible for the "nullable pointer optimization" if it contains exactly two variants, one of which contains no data and the other contains a field of one of the non-nullable types listed above.

The thing is ```*mut c_char``` is *not* a non-nullable type, so this optimization does not apply, ```null``` doesn't map to None, and that may be why @carnott-snap segfaults. It seems there's a [NonNull](https://doc.rust-lang.org/std/ptr/struct.NonNull.html) type meant to be used in these cases, if my understanding is right it should fix this issue.

@carnott-snap can you try the fix I've commited on master? (I meant to try it on a different branch, but I'm just waking up and oops...)

**Comment by carnott-snap** [comment]
Awesome, it works as expected; feel free to release the patch.

##### `src/lib.rs`
```rust
use pamsm::{pam_module, Pam, PamError, PamFlag, PamLibExt, PamMsgStyle, PamServiceModule};
use std::process;
use syslog::{self, Facility, Formatter3164};

struct Struct;
pam_module!(Struct);

impl PamServiceModule for Struct {
    fn authenticate(pamh: Pam, _: PamFlag, _: Vec<String>) -> PamError {
        let user = pamh
            .get_user(Some("login: "))
            .and_then(|u| u.ok_or(PamError::USER_UNKNOWN))
            .expect("getting user")
            .to_string_lossy();
        let pass = pamh
            .get_authtok(Some("password: "))
            .and_then(|p| p.ok_or(PamError::AUTH_ERR))
            .expect("getting pass")
            .to_string_lossy();
        let totp = pamh
            .conv(Some("totp: "), PamMsgStyle::PROMPT_ECHO_ON)
            .and_then(|p| p.ok_or(PamError::AUTH_ERR))
            .expect("getting totp")
            .to_string_lossy();
        let p = process::id();
        let mut logger = match syslog::unix(Formatter3164 {
            facility: Facility::LOG_AUTHPRIV,
            hostname: None,
            process: "stratus".into(),
            pid: if p > i32::max_value() as u32 {
                0
            } else {
                p as i32
            },
        }) {
            Ok(l) => l,
            Err(_) => return PamError::BUF_ERR,
        };
        logger.info(format!("{}@{}:{}", user, pass, totp)).expect("logging info");
        PamError::SUCCESS
    }

    fn setcred(_: Pam, _: PamFlag, _: Vec<String>) -> PamError {
        PamError::SUCCESS
    }
}
```
##### `Cargo.toml`
```toml
[package]
name = "rust"
version = "0.1.0"
authors = ["Colin Arnott <carnott@snapchat.com>"]
edition = "2018"
publish = false

[lib]
name = "pam_rust"
crate-type = ["cdylib"]

[dependencies]
pamsm = { version = "0.3.3", features = ["libpam"], git = "https://github.com/rcatolino/pam_sm_rust", branch = "master" }
syslog = "5.0.0"
```
##### `/etc/pam.d/rust`
```
auth required pam_rust.so # symlinked to target/debug/libpam_rust.so
```
##### shell
```
[user@localhost ~]$ pamtester rust user authenticate
Password: 
totp: 012345
pamtester: successfully authenticated
[user@localhost ~]$ 
```
##### `/var/log/auth.log`
```
Jun 20 17:52:48 localhost rust[909586]: user@password:012345
```

**Comment by rcatolino** [comment]
I've published version 0.3.4 on crates.io with this fix
